### PR TITLE
S3 upload component

### PR DESCRIPTION
**Pull request by daniel-heppner-ibigroup**
- Adds a socket.io server that is run separately with `yarn server-dev`
- Adds new alert dropdown that shows if the user is disconnected from the server
- Adds disabled field functionality for when there is no socket connection

### CODE REVIEW

**Comment by daniel-heppner-ibigroup** [comment]
Need to type script the server file. Should we use something like `ts-node` or is there a better way? 
Also need to import the .env

File: pages/api/recordUpdate.ts [hunk 1]
```diff
@@ -1,5 +1,7 @@
 // Next.js API route support: https://nextjs.org/docs/api-routes/introduction
```
Comment (miles-grant-ibigroup) [inline pages/api/recordUpdate.ts]: We should be able to delete this file now, right?

File: package.json [hunk 2]
```diff
@@ -7,13 +7,18 @@
     "start": "next start",
     "lint": "next lint",
     "storybook": "start-storybook -p 6006",
-    "build-storybook": "build-storybook"
+    "build-storybook": "build-storybook",
+    "server-dev": "nodemon server/index.js"
```
Comment (miles-grant-ibigroup) [inline package.json]: Could we add a `server` script that just runs the server?

File: server/index.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
+const aws = require('aws-sdk')
+const { parse } = require('json2csv')
+const { Server } = require('socket.io')
+
+const PORT = process.env.PORT || 4000
+
+const userQuestionMatrix = {}
```
Comment (miles-grant-ibigroup) [inline server/index.js:7]: We might be able to typescript this and use our tsdx to package it, but I'll leave it up to the second reviewer to decide if this is necessary

File: README.md [hunk 4]
```diff
@@ -29,5 +29,8 @@ Open [http://localhost:3000](http://localhost:3000) with your browser to see the
 
 To see the localization support, append a language code at the end of the URL (for example, [http://localhost:3000/fr](http://localhost:3000/fr) will render French strings). The system will attempt to map that key to languages defined in the `next.config.js`. Language strings are stored in `config.yaml`.
 
+### How to get the S3 upload to work
+
+An `env.example` file is provided. By placing AWS credentials and a bucket name in the corresponding slots, next will pick up on them and automatically upload results into the specified bucket.
```
Comment (miles-grant-ibigroup) [inline README.md]: This may no longer be accurate

File: env.example [hunk 5]
```diff
@@ -0,0 +1,4 @@
+AWS_REGION=us-east-1
```
Comment (miles-grant-ibigroup) [inline env.example]: I would convert this file to json and then import it directly into the server

File: server/index.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    return
+  }
+
+  const csv = parse(data)
+  try {
+    const upload = await new aws.S3.ManagedUpload({
+      params: {
+        Body: csv,
+        Bucket: BUCKET_NAME,
+        Key: `survey-responses/${user}.csv`
```
Comment (miles-grant-ibigroup) [inline server/index.js]: Could we put the user csvs in a folder based on date? That way when it comes to zipping up the day's responses it'll be easier to merge and concatenate. If someone provides their response over 2 separate days then that will be reflected in the data -- we can generate monthly merges that combine those together.

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
This is a great start! Very clean architecture and works & looks great.

Just a few nitpicks here and there. Still have to test the s3 upload edge cases some more, but overall this looks like it works very well.

File: env.example (hunk 5 above)
Comment (daniel-heppner-ibigroup) [reply env.example]: I was also thinking that, but I wasn't sure how this was going to be deployed. If it's just an ec2 instance then I agree that a json/yml file is probably a bit better

File: server/index.js (hunk 6 above)
Comment (daniel-heppner-ibigroup) [reply server/index.js]: Behold, folder

File: package.json (hunk 2 above)
Comment (daniel-heppner-ibigroup) [reply package.json]: Are you saying you just want to rename this to `server`? Or should it do something different too?

File: server/index.js (hunk 3 above)
Comment (daniel-heppner-ibigroup) [reply server/index.js:7]: I don't think tsdx is really designed to run scripts, it's pretty much just for publishing npm packages it seems. I think something like this might work though. https://github.com/TypeStrong/ts-node

File: package.json (hunk 2 above)
Comment (miles-grant-ibigroup) [reply package.json]: I think there should be a script which runs the server without `nodemon`

**Review by miles-grant-ibigroup — APPROVED** [review]
We discussed some potential adjustments/expansions of this offline, but I would put those in a separate PR. This part looks great to me and I think is ready for merge!

File: package.json (hunk 2 above)
Comment (daniel-heppner-ibigroup) [reply package.json]: oh I see for actually running it in prod yes yes of course

**Review by philip-cline — APPROVED** [review]
Survey PRs. always a pleasure.
